plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSink.java

```diff
@@ -0,0 +1,249 @@
+        checkState(!committed, "already committed");
+        if (closed) {
+            return;
+        }
+
+        writers.computeIfAbsent(partitionId, this::createWriter);
+        synchronized (writers.get(partitionId)) {
+            writeToExchangeStorage(partitionId, Slices.wrappedIntArray(data.length()));
+            writeToExchangeStorage(partitionId, data);
+        }
+    }
+
+    private ExchangeStorageWriter createWriter(int partitionId)
+    {
+        URI outputPath = outputDirectory.resolve(partitionId + DATA_FILE_SUFFIX);
+        try {
+            return exchangeStorage.createExchangeStorageWriter(outputPath, secretKey);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private void writeToExchangeStorage(int partitionId, Slice slice)
+    {
+        int position = 0;
+        while (position < slice.length()) {
+            SliceOutput pendingBuffer = pendingBuffers.computeIfAbsent(partitionId, ignored -> {
+                try {
+                    return freeBuffers.take();
+                }
+                catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    throw new RuntimeException(e);
+                }
+            });
+            int writableBytes = min(pendingBuffer.writableBytes(), slice.length() - position);
+            pendingBuffer.writeBytes(slice.getBytes(position, writableBytes));
+            position += writableBytes;
+
+            flushIfNeeded(partitionId, false);
+        }
+    }
+
+    private void flushIfNeeded(int partitionId, boolean finished)
+    {
+        SliceOutput buffer = pendingBuffers.get(partitionId);
+        if (!buffer.isWritable() || finished) {
+            if (!buffer.isWritable()) {
+                pendingBuffers.remove(partitionId);
+            }
+            try {
+                writers.get(partitionId).write(buffer.slice()).addListener(() -> {
+                    buffer.reset();
+                    freeBuffers.add(buffer);
+                    synchronized (this) {
+                        if (!blockedFuture.isDone()) {
+                            blockedFuture.complete(null);
+                        }
+                    }
+                }, directExecutor());
+            }
+            catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+    }
+
+    @Override
+    public long getSystemMemoryUsage()
+    {
+        return INSTANCE_SIZE + (long) numBuffers * writeBufferSize;
+    }
+
+    @Override
+    public void finish()
+    {
+        if (closed) {
+            return;
+        }
+        try {
+            for (Integer partitionId : writers.keySet()) {
+                flushIfNeeded(partitionId, true);
+                try {
+                    pendingBuffers.get(partitionId).close();
```

Comment:
The buffers is just a memory, they don't have to be closed. (here and in other places). It should be sufficient to cleanup the containers to release the memory